### DISCUSSION TITLE
route with path parameters and : returns 422

### DISCUSSION THREAD

**Discussion by SamaritanK** [fastapi/fastapi#9702] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.get("/songs/{id}:play")
def reprocess_pill(id: int):
    return {"done"}
```

### Description

I read already https://github.com/tiangolo/fastapi/discussions/8513 but this problem is different.

dependencies:
fastapi 0.97.0
starlette 0.27.0

the url http://localhost:8000/v1/songs/7:play
returns 422 with a validation error
```
{
  "detail": [
    {
    "loc": [
      "path",
      "id"
    ],
    "msg": "value is not a valid integer",
    "type": "type_error.integer"
    }
  ]
}
```

### Operating System

macOS

### FastAPI Version

fastapi 0.97.0 starlette 0.27.0

### Python Version

3.10.9

**Comment by mojtabapaso** [comment]
Hello
I have never seen : used in the path parameter, it is better to use / for spacing.

@router.get("/songs/{id}/play")
def reprocess_pill(id: int):
    return {"done"}

**Comment by jgould22** [comment]
https://www.starlette.io/routing/

```
 By default this will capture characters up to the end of the path or the next /.
```

By default the extractor is probably grabbing the entire bit `7:play` and then checking if it is an int, which it will not be.

You can probably check by changing the value to a string temporarily and see what it is grabbing.

**Comment by SamaritanK** [comment, marked as answer]
Thanks for your suggestions, I solved it.
It's a matter of routes definition sequence (unluckily the order matters) , I have 2 routes:

```
@router.get("/songs/{id}")
def get( id: int ):
...

@router.get("/songs/{id}:play")
def play(id: int):
...
```

calling http://localhost:8000/v1/songs/7:play the first is triggered with 7:play as value => validation error
Changing the order fixed it
```
@router.get("/songs/{id}:play")
def play(id: int):
...

@router.get("/songs/{id}")
def get( id: int ):
...
```

We are using ':' for custom methods where the API is not expressing a REST access
https://cloud.google.com/apis/design/custom_methods here an example
